### DISCUSSION TITLE
[@nuxt/kit] Can't use `addPluginTemplate` after `pages:extend` hook in module

### DISCUSSION THREAD

**Discussion by victorgarciaesgi** [nuxt/nuxt#16865] [Questions]
### Environment

- Operating System: `Darwin`
- Node Version:     `v16.15.0`
- Nuxt Version:     `3.0.0-rc.9`
- Nitro Version:    `0.5.0`
- Package Manager:  `yarn@1.22.18`
- Builder:          `vite`
- User Config:      `-`
- Runtime Modules:  `-`
- Build Modules:    `-`

### Reproduction

https://github.com/victorgarciaesgi/nuxt-typed-router/tree/bug/addPlugin

You can checkout the branch `bug/addPlugin`

`yarn build:local && yarn play`

### Describe the bug

In my module `nuxt-typed-router`, I want to create a runtime plugin. But to create this plugin, I depend on pages getting resolved. [I use `nuxt.hook('pages:extend')` to have all the routes tree](https://github.com/victorgarciaesgi/nuxt-typed-router/blob/bug/addPlugin/src/module.ts).

Inside this hook, I can get routes names from it and generate a declaration tree, and [use it in my plugin template](https://github.com/victorgarciaesgi/nuxt-typed-router/blob/bug/addPlugin/src/generators/nuxtHook.ts)

I added `write: true` to see if the plugin is generated and there is nothing in the `.nuxt` folder. I tried outside the hook and it is correctly generated, but I can't access the info I need outside the hook.

I also tried the `runtimeConfig` approach like in https://stackblitz.com/edit/github-jauvhr?file=modules%2Fhello.ts , but with no success

**Comment by tuarrep** [comment]
Hello,

Same here `components:extend` hook.

Here the link to my WIP module : https://github.com/tuarrep/nuxt-storm/commit/fd0d3f1515e1f20370b13db24850dca9740abe69

If it can helps, here what `addTemplate` returns when I log it (the folder in paths differ because of `yarn link`) :
```js
{                                                                                                                                                                                                                                                            nuxt-storm 18:13:58
  src: '/Users/nicolas/Dev/nuxt-storm/templates/components.js',
  fileName: '/Users/nicolas/Dev/docgen/.components.gen.js',
  options: {
    components: [
      [Object],
      // ... repeated
    ]
  },
  write: true,
  filename: '/Users/nicolas/Dev/docgen/.components.gen.js',
  dst: '/Users/nicolas/Dev/docgen/.components.gen.js'
}

```

**Comment by danielroe** [comment]
You should not be adding templates within hooks, for exactly this reason. Those functions are meant to be called at the time of your module setup function.

**Reply by victorgarciaesgi** [reply]
Is there an other way to access the generated pages array then? I didn't find any. In Nuxt 2 we could add a plugin after the pages were resolved.
For now to make it work I have to add it manually to the `plugins` folder

**Reply by tuarrep** [reply]
That was exactly what the module does for Nuxt 2. See the main branch.

What is the recommended way to do this for Nuxt 3?

**Reply by victorgarciaesgi** [reply]
Maybe a `pages:resolved` hook that is called before `build:before`? That way we can access the generated pages tree and add a plugin that depends on it?

**Reply by victorgarciaesgi** [reply]
That's the workaround I use on the master branch: https://github.com/victorgarciaesgi/nuxt-typed-router/blob/master/src/generators/nuxtHook.ts

It's not clean but that was the only way. 

```ts
nuxt.hook('build:done', async () => {
    const pluginFolder = `${srcDir}/plugins`;
    await saveRouteFiles(
      pluginFolder,
      srcDir,
      pluginName,
      createRuntimePluginFile(routesDeclTemplate)
    );
  });
```

**Reply by danielroe** [reply]
You can add a virtual template with a `getContents` function, which can then refer to data that you set in a hook. See some of the ad-hoc modules within packages/nuxt in this repository for examples.

**Reply by tuarrep** [reply]
I ended up using `writeFile` from `fs`
